server/boards/app/files.go

```diff
@@ -175,3 +190,150 @@ func (a *App) MoveFile(channelID, teamID, boardID, filename string) error {
+	for _, block := range blocks {
+		if block.Type == model.TypeImage || block.Type == model.TypeAttachment {
+			fieldName := "fileId"
+			if fileID, ok := block.Fields[fieldName]; ok {
+				blockIDs = append(blockIDs, block.ID)
+				blockPatches = append(blockPatches, model.BlockPatch{
+					UpdatedFields: map[string]interface{}{
+						fieldName: fileID,
+					},
+					DeletedFields: []string{"attachmentId"},
+				})
+			}
+		}
+	}
+	a.logger.Debug("Duplicate boards patching file IDs", mlog.Int("count", len(blockIDs)))
+
+	if len(blockIDs) != 0 {
+		patches := &model.BlockPatchBatch{
+			BlockIDs:     blockIDs,
+			BlockPatches: blockPatches,
+		}
+		if err := a.store.PatchBlocks(patches, userID); err != nil {
+			return fmt.Errorf("could not patch file IDs while duplicating board %s: %w", boardID, err)
+		}
+	}
+
+	return nil
+}
+
+func (a *App) CopyCardFiles(sourceBoardID string, copiedBlocks []*model.Block, asTemplate bool) error {
+	// Images attached in cards have a path comprising the card's board ID.
+	// When we create a template from this board, we need to copy the files
+	// with the new board ID in path.
+	// Not doing so causing images in templates (and boards created from this
+	// template) to fail to load.
+
+	// look up ID of source sourceBoard, which may be different than the blocks.
+	sourceBoard, err := a.GetBoard(sourceBoardID)
+	if err != nil || sourceBoard == nil {
+		return fmt.Errorf("cannot fetch source board %s for CopyCardFiles: %w", sourceBoardID, err)
+	}
+
+	var destBoard *model.Board
+
+	for _, block := range copiedBlocks {
+		if block.Type != model.TypeImage && block.Type != model.TypeAttachment {
+			continue
+		}
+
+		fileId, isOk := block.Fields["fileId"].(string)
+		if !isOk {
+			fileId, isOk = block.Fields["attachmentId"].(string)
+			if !isOk {
+				continue
+			}
+		}
+
+		// create unique filename
+		ext := filepath.Ext(fileId)
+		fileInfoID := utils.NewID(utils.IDTypeNone)
+		destFilename := fileInfoID + ext
+
+		if destBoard == nil || block.BoardID != destBoard.ID {
+			destBoard = sourceBoard
+			if block.BoardID != destBoard.ID {
+				destBoard, err = a.GetBoard(block.BoardID)
+				if err != nil {
+					return fmt.Errorf("cannot fetch destination board %s for CopyCardFiles: %w", sourceBoardID, err)
+				}
+			}
+		}
+
+		// GetFilePath will retrieve the correct path
+		// depending on whether FileInfo table is used for the file.
+		fileInfo, sourceFilePath, err := a.GetFilePath(sourceBoard.TeamID, sourceBoard.ID, fileId)
+		if err != nil {
+			return fmt.Errorf("cannot fetch destination board %s for CopyCardFiles: %w", sourceBoardID, err)
+		}
+
+		destinationFilePath := filepath.Join(utils.GetBaseFilePath(), destFilename)
+		// Global Templates are handled via Import, if user-defined templates
+		// are to be stored by team. Won't be deleted by Data Retention
+		if asTemplate {
+			destinationFilePath = filepath.Join(destBoard.TeamID, destBoard.ID, destFilename)
```

Comment:
All of this magic path joining and field fetching should really be done in once place. Otherwise it's always going to desync.

Reply:
Yep, move these to a common function.